tests/conftest.py

```diff
@@ -9,8 +9,12 @@
 from asgi_lifespan import LifespanManager
 from sqlalchemy_utils import create_database, database_exists, drop_database
 
+from server.application.auth.commands import CreateUser
+from server.application.auth.queries import GetUserByEmail
 from server.config import Settings
 from server.config.di import bootstrap, resolve
+from server.domain.auth.entities import User, UserRole
+from server.seedwork.application.messages import MessageBus
```

Comment:
Can you at some point fill me in with some general info on the dependency injection system, and where/how is `MessageBus` implemented? I see that it's `raise NotImplementedError` in `server/seedwork/application/messages.py`

Reply:
Alors...

La pièce principale de l'injection de dépendance est `server/config/di.py` : c'est ici qu'on configure le "DI container", qui va faire le lien entre ce qu'on passe à `resolve()` et ce qui doit être ainsi résolu.

Si j'y définis :

```python
container.register(MyClass, instance=MySpecificClassImplementation())
```

Et que j'appelle `resolve(MyClass)`, alors j'obtiendrai l'instance `MySpecificClassImplementation`.

Dans cet exemple comme dans celui de `MessageBus`, `MyClass` est une "interface" (au sens de la POO) que l'implémentation implémente. Donc oui, il est attendu que `MessageBus` soit une classe "vide" (une interface, en réalité), et on a le `MessageBusAdapter` qui l'implémente. Quand le code appelle `resolve(MessageBus)`, il obtient en fait typiquement l'instance du `MessageBusAdapter`.

Dans le code on ne "voit" que l'interface, donc en effet ce n'est pas très pratique si on veut aller "voir ce que le `MessageBus`" fait. Il faut juste se rappeler qu'il y a une implémentation, et que pour voir laquelle c'est on peut aller dans `di.py` voir comment on "relie tout ça ensemble".

Le but de ce découplage par interfaces est (théoriquement au moins) de pouvoir "swap" les implémentations "par défaut" lors des tests. Tu verras qu'on le fait déjà dans `tests/domain/test_datasets.py`, où on remplace la résolution du `DatasetRepository` par un `MockDatasetRepository` qui ne passe pas par la DB.